docs: Voting ledger design (#709)
input-output-hk/catalyst-voices: docs/src/architecture/08_concepts/immutable_ledger/ledger.md

# Immutable Ledger Structure

---

Title: Immutable Ledger Structure Design

Status: Proposed

Authors:
    - Alex Pozhylenkov <alex.pozhylenkov@iohk.io>

Created: 2024-08-19

---

## Abstract

This document describes a specification of the immutable ledger structure for various purposes of project "Catalyst".

## Motivation

Project "Catalyst" requires a solution for storing people votes and any other data,
in a transparent, verifiable, scalable and immutable way.

## Specification

### Ledger structure

![Ledger schema](images/ledger_schema.svg){ align=right }

> Ledger will be represented as a collection of different, non-connected chains,

Reviewer: ```suggestion
Ledger will be represented as a collection of distinct, unconnected chains,
```

processed and run in parallel.
The only common thing for all these chains will be a "tree" identifier,
so these chains will serve and form an overall ledger state.

Obviously, given approach leads to data duplication,
as each chain, will not know anything about others.
And it also requires that the overall ledger state,
could be deterministically defined at any point of time,
considering potential transaction overlapping or duplication.

> Each particular chain, will be as a common sequence of blocks,

Reviewer: ```suggestion
Each particular chain, will be a common sequence of blocks,
```

which are cryptographically protected by hashing.

The described approach allows to easily scale and increase throughput of the network on demand at any time,
just by starting to process new chains.
<!-- markdownlint-disable no-inline-html -->
<br clear="right"/>
<!-- markdownlint-enable no-inline-html -->

### Temporary chains

![Temporary chain schema](images/temporary_chain.svg){ align=right }

It is a common thing for blockchains to have a starting block (genesis),

> but it's unusual to have a final block for a chain.

Reviewer: A bit picky,  but for consistency in the document, either one of  "it is" or "it's" should be used for the entire document.

> After which one no any block could be produced.

Reviewer: ```suggestion
After which no any block could be produced.
```
